File: le-liaison/src/main/java/com/latticeengines/liaison/exposed/service/Query.java
package com.latticeengines.liaison.exposed.service;

import com.latticeengines.liaison.exposed.exception.DefinitionException;
import com.latticeengines.liaison.exposed.service.QueryColumn;
import java.lang.RuntimeException;
import java.util.ArrayList;
import java.util.List;

public abstract class Query {
	
	private String name;
	private List<QueryColumn> columns;
	private List<String> columnNames;
	private List<String> filterDefns;
	private List<String> entityDefns;
	private String resultSetDefn;
	
	public Query( String name, String definition ) throws DefinitionException {
		initFromDefinition( name, definition );
	}
	
	public Query( String name, List<QueryColumn> columns, List<String> filterDefns, List<String> entityDefns, String resultSetDefn ) {
		initFromValues( name, columns, filterDefns, entityDefns, resultSetDefn );
	}
	
	public Query( Query other ) {
		initFromValues( other.name, other.columns, other.filterDefns, other.entityDefns, other.resultSetDefn );
	}
	
	public String getName() {
		return name;
	}
	
	public void setName( String name ) {
		this.name = name;
	}
	
	public List<QueryColumn> getColumns() {
		return columns;
	}
	
	public List<String> getColumnNames() {
		return columnNames;
	}
	
	public List<String> getFilterDefns() {
		return filterDefns;
	}
	
	public List<String> getEntityDefns() {
		return entityDefns;
	}
	
	public String getResultSetDefn() {
		return resultSetDefn;
	}
	
	public QueryColumn getColumn( String colname ) throws RuntimeException {
		if( !columnNames.contains(colname) ) {
			throw new RuntimeException( String.format("Column \"%s\" does not exist in query \"%s\"",colname,getName()) );
		}
		int idx = columnNames.indexOf( colname );
		return columns.get( idx );
	}
	
	public void appendColumn( QueryColumn column ) {
		if( !columnNames.contains(column.getName()) ) {
			columns.add( createColumn(column) );
			columnNames.add( column.getName() );
		}
	}
	
	public void removeColumn( String colname ) {
		if( columnNames.contains(colname) ) {
			int idx = columnNames.indexOf( colname );
			columnNames.remove( idx );
			columns.remove( idx );
		}
	}
	
	public void updateColumn( QueryColumn column ) {
		if( columnNames.contains(column.getName()) ) {
			int idx = columnNames.indexOf( column.getName() );
			columns.get( idx ).initFromValues( column.getName(), column.getExpression(), column.getMetadata() );
		}
		else {
			appendColumn( column );
		}
	}
	
	public void initFromValues( String name, List<QueryColumn> columns, List<String> filterDefns, List<String> entityDefns, String resultSetDefn ) {
		this.name = name;
		this.columns = new ArrayList<>();
		this.columnNames = new ArrayList<>();
		this.filterDefns = new ArrayList<>(filterDefns);
		this.entityDefns = new ArrayList<>(entityDefns);
		this.resultSetDefn = resultSetDefn;
		for( QueryColumn qc : columns ) {
			this.columns.add( createColumn(qc) );
			this.columnNames.add( qc.getName() );
		}
	}
	
	public abstract QueryColumn createColumn( QueryColumn qc );
	
	public abstract void initFromDefinition( String name, String defn ) throws DefinitionException;
	
	public abstract String definition();
	
}

File: le-liaison/src/main/python/liaison/Query.py

#
# $LastChangedBy$
# $LastChangedDate$
# $Rev$
#

class Query(object):

    def __init__(self, name, columns = [], filters = [], entities = [], resultSet = None):

        self.initFromValues(name, columns, filters, entities, resultSet)


    def getName(self):
        return self.name_

    def setName(self, n):
        self.name_ = n

    def getColumns(self):
        return self.columns_

    def getColumnNames(self):
        return self.column_names_

    def setColumns(self, cc):
        self.columns_ = cc
        self.column_names_ = []
        for c in cc:
            self.column_names_.append(c.getName())

    def getColumn(self, colname):
        if colname in self.column_names_:
            idx = self.column_names_.index(colname)
            return self.columns_[idx]
        return None

    def appendColumn(self, c):
        if c.getName() not in self.column_names_:
            self.columns_.append(c)
            self.column_names_.append(c.getName())

    def removeColumn(self, colname):
        if colname in self.column_names_:
            self.column_names_.remove(colname)
            c = (c for c in self.columns_ if c.getName() == colname).next()
            self.columns_.remove(c)

    def updateColumn(self, c):
        if c.getName() in self.column_names_:
            idx = self.column_names_.index(c.getName())
            self.columns_[idx] = c
        else:
            self.appendColumn(c)

    def getFilters(self):
        return self.filters_

    def setFilters(self, ff):
        self.filters_ = ff

    def getEntities(self):
        return self.entities_

    def setEntities(self, ee):
        self.entities_ = ee

    def initFromValues(self, name, columns, filters, entities, resultSet):

        self.name_         = name
        self.columns_      = columns
        self.column_names_ = []
        for c in columns:
            self.column_names_.append( c.getName() )
        self.filters_      = filters
        self.entities_     = entities
        self.resultSet_    = resultSet
